basic_pipeline/8_Mixer.md

```diff
@@ -0,0 +1,232 @@
+A careful reader might notice, that we are holding only a reference to only one buffer for each track, instead of a list of all the potentially unprocessed buffers - does it mean that we are losing some of them? Not at all, since we are taking advantage of the elements which have appeared earlier in the pipeline and which provide us an ordered list of frames on each of the inputs - however, we will need to process each buffer just at the moment it comes on the pad.
+
+The logic we gonna implement can be described in the following three steps:
++ If all the 'active' tracks (in the `:started` state or the ones in the `:finished` state but with the unprocessed buffer in the `Track` structure) - output the one with the lower timestamp. Otherwise do nothing.
++ If all the tracks are in the `:finished` state and their `:buffer` is empty - send the `:end_of_stream` event.
++ For all the tracks which are in the `:started` state and their buffer is empty - demand on the pad corresponding to that track.
+
+The next step in our element implementation is quite an obvious one:
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ @impl true
+ def handle_init(_options) do
+  {:ok,
+  %{
+    tracks: %{first_input: %Track{}, second_input: %Track{}}
+  }}
+ end
+ ...
+end
+```
+We have provided an `handle_init/1` callback, which does not expect any options to be passed. We are simply setting up the structure of the element state.
+As mentioned previously, we will have a `Track` structure for each of the input pads.
+Following on the callbacks implementation, let's continue with `handle_process/4` implementation:
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ @impl true
+ def handle_process(pad, buffer, _context, state) do
+  tracks =
+    Map.update!(state.tracks, pad, fn track ->
+      %Track{track | buffer: buffer}
+    end)
+
+  state = %{state | tracks: tracks}
+  {{:ok, [{:redemand, :output}]}, state}
+ end
+ ...
+end
+```
+What we do is that we are simply putting the incoming `buffer` into the `Track` structure for the given pad - note, that we are sure that we are not losing any information which is in the `Track`'s buffer before the update. Why can we be sure of that? That's because we will be returning the `:redemand` action. As you might remember from the chapter about the redemands, all the actions returned from the element's callback are taken immediately after the callback returns - that means, they are taken before any other incoming event is served. If we would have put the whole logic of processing the buffer into the `handle_demand/5` callback, we could simply call that callback with the `:redemand` action and be sure that the newly added buffer will be immediately processed before any other buffer comes (that means - before the other `handle_process/4` callback gets called).
```

Comment:
I was trying to point out, that due to the fact that :redemand action is performed synchroniously, right after the callback returns, we can be sure, that, for instance, there won't come any other buffer between this callback, and the handle_demand callback launched by :redemand action (note, that we are capable of holding at most 1 buffer for each track, since we have resigned from the buffer in form of a list). I will try to rewrite it to make it more clear

Reply:
done